src/Storages/ObjectStorage/DataLakes/DeltaLake/TableSnapshot.cpp

```diff
@@ -0,0 +1,262 @@
+        const DB::Names & partition_columns_,
+        DB::ObjectStoragePtr object_storage_,
+        LoggerPtr log_)
+        : scan(KernelUtils::unwrapResult(ffi::scan(snapshot_.get(), engine_.get(), /* predicate */{}), "scan"))
+        , scan_data_iterator(KernelUtils::unwrapResult(
+            ffi::kernel_scan_data_init(engine_.get(), scan.get()),
+            "kernel_scan_data_init"))
+        , data_prefix(data_prefix_)
+        , schema(schema_)
+        , partition_columns(partition_columns_)
+        , object_storage(object_storage_)
+        , log(log_)
+    {
+    }
+
+    size_t estimatedKeysCount() override
+    {
+        /// For now do the same as StorageObjectStorageSource::GlobIterator.
+        /// TODO: is it possible to do a precise estimation?
+        return std::numeric_limits<size_t>::max();
+    }
+
+    DB::ObjectInfoPtr next(size_t) override
+    {
+        std::lock_guard lock(next_mutex);
+        while (data_files.empty())
+        {
+            auto have_scan_data_res = KernelUtils::unwrapResult(
+                ffi::kernel_scan_data_next(scan_data_iterator.get(), this, visitData),
+                "kernel_scan_data_next");
+
+            if (!have_scan_data_res)
+                return nullptr;
+        }
+
+        chassert(!data_files.empty());
+
+        auto object = data_files.front();
+        data_files.pop_front();
+
+        chassert(object);
+        return object;
+    }
+
+    static void visitData(
+        void * engine_context,
+        ffi::ExclusiveEngineData * engine_data,
+        const struct ffi::KernelBoolSlice selection_vec,
+        const ffi::CTransforms *transforms)
+    {
+        ffi::visit_scan_data(engine_data, selection_vec, transforms, engine_context, Iterator::scanCallback);
+    }
+
+    static void scanCallback(
+        ffi::NullableCvoid engine_context,
+        struct ffi::KernelStringSlice path,
+        int64_t size,
+        const ffi::Stats * stats,
+        const ffi::DvInfo * /* dv_info */,
+        const struct ffi::CStringMap * partition_map)
+    {
+        auto * context = static_cast<TableSnapshot::Iterator *>(engine_context);
+        std::string full_path = fs::path(context->data_prefix) / KernelUtils::fromDeltaString(path);
+
+        /// Collect partition values info.
+        /// DeltaLake does not store partition values in the actual data files,
+        /// but instead in data files paths directory names.
+        /// So we extract these values here and put into `partitions_info`.
+        DB::ObjectInfoWithPartitionColumns::PartitionColumnsInfo partitions_info;
+        for (const auto & partition_column : context->partition_columns)
+        {
+            auto * raw_value = ffi::get_from_string_map(
+                partition_map,
+                KernelUtils::toDeltaString(partition_column),
+                KernelUtils::allocateString);
+
+            auto value = std::unique_ptr<std::string>(static_cast<std::string *>(raw_value));
```

Comment:
why not just to check `if (raw_value)` and than initialize std::string from this `char *`. I'm not sure it's safe to `static_cast` to `std::string *`